posts/non-fungible-tokens-on-the-ethereum-blockchain.md

```diff
@@ -0,0 +1,523 @@
+                myContext.tokenAddress = newAddress;
+
+            } else {
+                console.log('No contract deployed on network ' + netId);
+                myContext.tokenAddress = '';
+            }
+        }
+    });
+
+    await this.setState({web3ctx: {...myContext}});
+};
+```
+
+Once you get MetaMask installed and authenticated to a wallet, you should see the start page:
+
+![Home](https://raw.githubusercontent.com/misterzero/blog-usa/master/images/2019/01/non-fungible-tokens-04.png)
+
+The `Token` tab exposes settings used for the IPFS gateway and token contract. We're storing the unique image on the Interplanetary File System (IPFS) because the blockchain is not efficient for (or in some cases _capable of_) storing large binary files. You can read more about IPFS [on their site](https://ipfs.io/).
+
+![Token](https://raw.githubusercontent.com/misterzero/blog-usa/master/images/2019/01/non-fungible-tokens-05.png)
+
+Although IPFS is distributed, you can't access IPFS unless you're running a node, so we're pointing to IPFS Gateways to upload and access our files (more on this later). Since I've never heard of a website with 100% uptime, we encourage our users to verify the selected gateway is up:
+
+![IPFS](https://raw.githubusercontent.com/misterzero/blog-usa/master/images/2019/01/non-fungible-tokens-06.png)
+
+Now we get to the fun part: uploading our image and metadata to IPFS! Click the `Choose File` button for file picker dialog. Once you've chosen a file, the `Upload File` button is enabled.
+
+![Upload](https://raw.githubusercontent.com/misterzero/blog-usa/master/images/2019/01/non-fungible-tokens-07.png)
+
+Here's the code that pushes the uploaded image to our IPFS gateway:
+```javascript
+await ipfs.add(this.state.myToken.imageBuffer, (err, ipfsHash) => {
+    console.log('IPFS hash: ' + ipfsHash + ', error: ' + err);
+
+    let myMetadata = JSON.parse(JSON.stringify(this.state.myToken.metadataBuffer));
+    myMetadata.image = this.state.web3ctx.ipfsGateway + ipfsHash[0].hash;
+
+    this.setState({
+        myToken: {
+            ...this.state.myToken,
+            ipfsImageHash: ipfsHash[0].hash,
+            ipfsImageUrl: myMetadata.image,
+            metadataBuffer: myMetadata,
+            imageBuffer: ''
+        },
+        selectedTab: 'Metadata'
+    });
+})
+```
+
+After you've uploaded your file to IPFS, you should land on the `Metadata` tab with your IPFS file URI pre-filled in the metadata JSON input field:
+
+![Upload](https://raw.githubusercontent.com/misterzero/blog-usa/master/images/2019/01/non-fungible-tokens-08.png)
+
+Note in above that the IPFS image is displayed in the status block on the right. One of the amazing features of IPFS is that you _cannot_ upload duplicate images, because every image is addressed by the hash of the file, and guess what happens if you hash the same file more than once? That's right folks, you get **the same hash**.
```

Comment:
```suggestion
Note in above that the IPFS image is displayed in the status block on the right. One of the fantastic features of IPFS is that you _cannot_ upload duplicate images, because the hash of the file addresses every image, and guess what happens if you hash the same file more than once? That's right folks, you get **the same hash**.
```